### PR TITLE
Add MAC address randomization on module load for qcacld-2.0

### PR DESCRIPTION
**Pull request by hagau**
This adds MAC address randomization on module load for qcacld-2.0 with a sysctl switch for disabling defaulting to on.

### CODE REVIEW

File: drivers/staging/qcacld-2.0/CORE/HDD/src/wlan_hdd_main.c [hunk 1]
```diff
@@ -14070,12 +14073,22 @@ static int hdd_cnss_wlan_mac(hdd_context_t *hdd_ctx)
...
 	uint32_t mac_addr_size = VOS_MAC_ADDR_SIZE;
 	u8 *addr, *buf;
+	u8 addr_random[VOS_MAX_CONCURRENCY_PERSONA * VOS_MAC_ADDR_SIZE];
 	struct device *dev = hdd_ctx->parent_dev;
 	hdd_config_t *ini = hdd_ctx->cfg_ini;
 	tSirMacAddr customMacAddr;
 	VOS_STATUS status;
 
-	addr = vos_get_cnss_wlan_mac_buff(dev, &no_of_mac_addr);
+	if(sysctl_randomize_mac) {
```
Comment (thestinger) [inline drivers/staging/qcacld-2.0/CORE/HDD/src/wlan_hdd_main.c]: `if (`

File: drivers/staging/qcacld-2.0/CORE/HDD/src/wlan_hdd_main.c [hunk 2]
```diff
@@ -14070,12 +14073,22 @@ static int hdd_cnss_wlan_mac(hdd_context_t *hdd_ctx)
...
+		get_random_bytes(addr_random, sizeof(addr_random));
+		for (iter = 0; iter < VOS_MAX_CONCURRENCY_PERSONA * VOS_MAC_ADDR_SIZE; iter += VOS_MAC_ADDR_SIZE) {
+			addr_random[iter] &= (addr_random[iter] & 0xfe) | 2;
+		}
+		addr = (u8*) &addr_random;
```
Comment (thestinger) [inline drivers/staging/qcacld-2.0/CORE/HDD/src/wlan_hdd_main.c]: `(u8*)&addr_random`

File: net/core/sysctl_net_core.c [hunk 3]
```diff
@@ -92,6 +95,13 @@ static int rps_sock_flow_sysctl(ctl_table *table, int write,
 
 static struct ctl_table net_core_table[] = {
 #ifdef CONFIG_NET
+	{
+		.procname       = "randomize_mac",
```
Comment (thestinger) [inline net/core/sysctl_net_core.c]: If you make it `net.randomize_mac`, the system property needs to be updated to use that.

The Pixel implementation will need to be kept in sync with this implementation too. They need to function the same way and implement the same sysctl.

Comment (hagau) [reply net/core/sysctl_net_core.c]: sorry, this patch was written when it was still net.randomize_mac. I have moved it to /proc/sys/dev/cnss/randomize_mac.

**Comment by thestinger** [comment]
Thanks for helping with this. I ended up backporting the Pixel 2 implementation to the Pixel to replace the older implementation (requiring some porting) and it applied cleanly here so we won't need a special-cased implementation.

It's a bit different because it's leaving the vendor prefix intact since that also avoids having a unique MAC address but also avoids broadcasting that it's as a device randomizing the MAC address in a very obvious way. Similarly, it may make it harder to fingerprint it as a CopperheadOS device based on the fact that it's using MAC randomization and fits the profile of supported CopperheadOS hardware.

**Comment by thestinger** [comment]
Ah, the code actually needs to be moved for the 5X since it uses wlan_mac.bin. It's going to be the same concept as this though, to preserve the vendor prefix.

**Comment by thestinger** [comment]
*sigh* :\

**Comment by thestinger** [comment]
The new qcacld-2.0 Pixel code is https://github.com/CopperheadOS/kernel_google_marlin/commit/669b5add391ea6e94934519f12adf7049ba7a6eb but the 5X uses the hdd_update_mac_config call in drivers/staging/qcacld-2.0/CORE/HDD/src/wlan_hdd_cfg.c rather than CNSS to retrieve the MAC(s).

**Comment by hagau** [comment]
I'm closing this since there is a better implementation merged.

**Comment by thestinger** [comment]
I don't think the new implementation works for the 5X approach to this though.

**Pull request closed without merging**